File: 哈希表/535. TinyURL 的加密与解密/535. TinyURL 的加密与解密.go
type Codec struct {
	dataBase map[int]string
    id int
}


func Constructor() Codec {
	return Codec{map[int]string{}, 0}
}

// Encodes a URL to a shortened URL.
func (this *Codec) encode(longUrl string) string {
	this.id++
	this.dataBase[this.id] = longUrl
	return "http://tinyurl.com/" + strconv.Itoa(this.id)
}

// Decodes a shortened URL to its original URL.
func (this *Codec) decode(shortUrl string) string {
    i := strings.LastIndexByte(shortUrl,'/')
	id, _ := strconv.Atoi(shortUrl[i+1:])
	return this.dataBase[id]
}


/**
 * Your Codec object will be instantiated and called as such:
 * obj := Constructor();
 * url := obj.encode(longUrl);
 * ans := obj.decode(url);
 */

File: 哈希表/535. TinyURL 的加密与解密/535. TinyURL 的加密与解密.js
/**
 * Encodes a URL to a shortened URL.
 *
 * @param {string} longUrl
 * @return {string}
 */

 var encode = function(longUrl) {
    this.dataBase = new Map();
    this.id = 0;
    this.id ++;
    this.dataBase.set(this.id,longUrl);
    return "http://tinyurl.com/" + this.id;
};
/**
 * Decodes a shortened URL to its original URL.
 *
 * @param {string} shortUrl
 * @return {string}
 */
var decode = function(shortUrl) {
    const p = shortUrl.lastIndexOf('/') + 1;
    const key = parseInt(shortUrl.substring(p));
    return this.dataBase.get(key);
}

/**
 * Your functions will be called as such:
 * decode(encode(url));
 */